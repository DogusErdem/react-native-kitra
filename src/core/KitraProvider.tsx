import type { ComponentType, ForwardedRef } from 'react';
import React from 'react';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { StyleProp, ViewStyle } from 'react-native';

import { ThemeProvider } from './theme/theme';
import { TypographyContext, TypographyProvider } from './typography/typography';
import { NotificationContextType, NotificationProvider } from './notification/notification';

const Feather = require('react-native-vector-icons/Feather').default;
const Ionicons = require('react-native-vector-icons/Ionicons').default;
const AntDesign = require('react-native-vector-icons/AntDesign').default;

Animated.addWhitelistedNativeProps({ text: true });
const messageTypes = (theme:any) => ({
  SUCCESS: {
    backgroundColor: theme.colors.status.successLight,
    icon: <Feather name="check" size={25} color={theme.colors.status.success} />,
  },
  INFO: {
    backgroundColor: theme.colors.system.primary15,
    icon: <Ionicons name="information-circle-outline" size={25} color={theme.colors.system.primary} />,
  },
  DANGER: {
    backgroundColor: theme.colors.status.errorLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.error} />,
  },
  WARNING: {
    backgroundColor: theme.colors.status.warningLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.warning} />,
  },
});

type KitraProviderType= {
  children?:any,
  limit?:number,
  notificationRef?:React.Ref<NotificationContextType>,
  messageType?:(theme?:any, colorScheme?:'dark' | 'light') => {[key:string]: {
      backgroundColor:string,
    icon?:React.ReactNode
    onPress?:()=>void
  }},
  notificationCcontainerStyle?:StyleProp<ViewStyle>,
  customView?:({ header, type, message, theme }:
    {header:string, type:string, message:string, theme:any})=>React.ReactNode}

export const KitraProvider =
({ children, messageType = messageTypes, notificationCcontainerStyle, limit, customView,
  notificationRef }: KitraProviderType) => (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <ThemeProvider>
        <TypographyProvider>
          <NotificationProvider
            messageType={theme => (messageType ? messageType(theme) : theme)}
            notificationCcontainerStyle={notificationCcontainerStyle}
            limit={limit}
            customView={item => customView?.(item)}
            ref={notificationRef}
          >
            {children}
          </NotificationProvider>
        </TypographyProvider>
      </ThemeProvider>
    </GestureHandlerRootView>
);

export function applyDefaults<T>(Component: ComponentType<T>) {
  return React.forwardRef((props: T, ref: ForwardedRef<any>) => (
    <TypographyContext.Consumer>
      {typography =>
        <Component ref={ref} typography={typography?.typography} {...props} />
          }
    </TypographyContext.Consumer>
  ));
}
