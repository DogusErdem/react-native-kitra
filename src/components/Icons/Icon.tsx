import React from 'react';
import { getIconType } from '../../utilities';
import { applyDefaults } from '../../core/KitraProvider';
import type { FCCWD, IconComponentProps, IconType } from '../../types';

type IconProps = {
    type: IconType,
} & IconComponentProps;

const Icon:FCCWD<IconProps> = ({ testID, type, ...props }) => {
  const IconComponent = getIconType(type || 'ionicons');

  return (
    <IconComponent testID={testID} {...props} />
  );
};

export default applyDefaults(Icon);
