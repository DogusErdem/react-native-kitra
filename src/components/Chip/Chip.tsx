import { TouchableOpacity, View } from 'react-native';
import { useEffect, useState } from 'react';
import Animated, { interpolateColor, useAnimatedStyle, useSharedValue, withTiming } from 'react-native-reanimated';
import useComponentTheme from '../../core/hooks/useComponentTheme';
import type { ChipProps, FCCWD } from '../../types';
import { applyDefaults } from '../../core/KitraProvider';
import Icon from '../Icons/Icon';

const AnimatedTouchableOpacity = Animated.createAnimatedComponent(TouchableOpacity);

const Chip:FCCWD<ChipProps> = (
  { typography,
    label,
    value,
    textStyle,
    style,
    icon,
    size = 'small',
    onChange,
    disabled = false,
    theme },
) => {
  const [select, setSelect] = useState(value || false);
  const componentStatus = disabled ? 'disabled' : (select ? 'active' : 'default');
  const { statusTheme } = useComponentTheme(theme, 'chip', componentStatus);

  const offset = useSharedValue(0);

  const typographySize = {
    small: typography?.body.xsmedium,
    medium: typography?.body.smedium,
    large: typography?.body.medium,
  };

  const animated = useAnimatedStyle(() => {
    const colorAnimated = interpolateColor(
      offset.value,
      [0, 1],
      [statusTheme.background, statusTheme.background],
    );
    return {
      backgroundColor: colorAnimated,
    };
  });

  useEffect(() => {
    offset.value = withTiming(Number(!!select));
    if (onChange) onChange(select);
  }, [select]);

  useEffect(() => {
    setSelect(!!value);
  }, [value]);
  return (
    <AnimatedTouchableOpacity
      disabled={disabled}
      testID="chip"
      onPress={() => { setSelect(prev => !prev); }}
      activeOpacity={1}
      style={[animated, { justifyContent: 'center',
        alignItems: 'center',
        height: sizes[size].height,
        borderWidth: 1,
        borderRadius: 40,
        opacity: disabled ? 0.7 : 1,
        width: icon ? sizes[size].iconwithWidth : sizes[size].width }, style, { borderColor: statusTheme.border }]}
    >
      <View style={{ flexDirection: icon?.iconPosition === 'right' ? 'row' : 'row-reverse', alignItems: 'center' }}>
        <Animated.Text style={[{ fontSize: typographySize[size]?.fontSize, lineHeight: typographySize[size]?.lineHeight, fontWeight: '500' }, textStyle, { color: statusTheme.label }]}>{label}</Animated.Text>
        {icon ?
          <Icon color={statusTheme.icon} type={icon.iconType} name={icon.iconName} size={sizes[size].iconSize} style={[icon.iconPosition === 'right' ? { marginLeft: 6 } : { marginRight: 6 }]} />
          : null
        }
      </View>
    </AnimatedTouchableOpacity>
  );
};

export default applyDefaults(Chip);

const sizes = {
  small: {
    width: 57,
    height: 24,
    iconSize: 12,
    iconwithWidth: 71.54,
  },
  medium: {
    width: 65,
    height: 30,
    iconSize: 15,
    iconwithWidth: 82.39,
  },
  large: {
    width: 73,
    height: 33,
    iconSize: 17,
    iconwithWidth: 93.24,

  },
};
