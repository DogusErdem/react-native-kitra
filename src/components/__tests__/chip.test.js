import { render, fireEvent } from '@testing-library/react-native';
import { Text, View } from 'react-native';
import { KitraProvider } from '../../core/KitraProvider';
import Chip from '../Chip/Chip';
import Icon from '../Icons/Icon';

/* eslint-disable no-undef */

const renderWithContext = children => render(<KitraProvider>{children}</KitraProvider>);

describe('Chip', () => {
  it('default render', () => {
    const tree = renderWithContext(<Chip label="Test" />);
    expect(tree).toMatchSnapshot();
  });
  describe('render with different size', () => {
    it('render with small size', () => {
      const { getByTestId } = renderWithContext(
        <Chip
          size="small"
          label="Test"
        />,
      );
      const container = getByTestId('chip');
      const label = getByTestId('chipLabel');
      expect(container).toHaveStyle({ height: 24 });
      expect(label).toHaveStyle({ fontSize: 10, lineHeight: 16 });
    });
    it('render with medium size', () => {
      const { getByTestId } = renderWithContext(
        <Chip
          size="medium"
          label="Test"
        />,
      );
      const container = getByTestId('chip');
      const label = getByTestId('chipLabel');
      expect(container).toHaveStyle({ height: 30 });
      expect(label).toHaveStyle({ fontSize: 12, lineHeight: 18 });
    });
    it('render with large size', () => {
      const { getByTestId } = renderWithContext(
        <Chip
          size="large"
          label="Test"
        />,
      );
      const container = getByTestId('chip');
      const label = getByTestId('chipLabel');
      expect(container).toHaveStyle({ height: 33 });
      expect(label).toHaveStyle({ fontSize: 14, lineHeight: 21 });
    });
  });
  describe('render with custom styles', () => {
    it('render with syle and labelStyle props', () => {
      const { getByTestId } = renderWithContext(
        <Chip
          style={{ padding: 5, marginTop: 10, paddingHorizontal: 3 }}
          labelStyle={{ fontSize: 14, lineHeight: 13, textAlign: 'center' }}
        />,
      );
      const container = getByTestId('chip');
      const label = getByTestId('chipLabel');
      expect(container).toHaveStyle({ padding: 5, marginTop: 10, paddingHorizontal: 3 });
      expect(label).toHaveStyle({ fontSize: 14, lineHeight: 13, textAlign: 'center' });
    });
  });
  describe('render with onChange and left right props', () => {
    const onChange = jest.fn();
    const left = jest.fn(item => <Text>{item ? 'LeftTrue' : 'LeftFalse'}</Text>);
    const right = jest.fn(item => <Text>{item ? 'RightTrue' : 'RightFalse'}</Text>);

    const { getByTestId, getByText } = renderWithContext(
      <Chip
        onChange={onChange}
        left={left}
        right={right}
      />,
    );
    const chip = getByTestId('chip');
    fireEvent.press(chip);
    expect(onChange).toHaveBeenCalledWith(true);
    expect(left).toHaveBeenCalledWith(true);
    expect(right).toHaveBeenCalledWith(true);
    expect(getByText('LeftTrue')).toBeTruthy();
    expect(getByText('RightTrue')).toBeTruthy();
    fireEvent.press(chip);
    expect(onChange).toHaveBeenCalledWith(false);
    expect(left).toHaveBeenCalledWith(false);
    expect(right).toHaveBeenCalledWith(false);
    expect(getByText('LeftFalse')).toBeTruthy();
    expect(getByText('RightFalse')).toBeTruthy();
  });
  describe('render with label prop', () => {
    it('label prop', () => {
      const { getByText } = renderWithContext(
        <Chip label="Chip Label Test" />,
      );
      expect(getByText('Chip Label Test')).toBeTruthy();
    });
  });
  describe('render with disabled prop', () => {
    it('disabled prop', () => {
      const onChange = jest.fn();
      const { getByTestId } = renderWithContext(
        <Chip
          onChange={onChange}
          disabled
        />,
      );
      const chip = getByTestId('chip');
      fireEvent.press(chip);
      fireEvent.press(chip);
      expect(onChange).toHaveBeenCalledTimes(1);
    });
  });
});

// REANİMATED İLE STİLLERİ KONTROL ET
