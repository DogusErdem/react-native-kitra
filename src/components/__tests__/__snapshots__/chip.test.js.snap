// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip default render 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    accessibilityState={
      {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "alignSelf": "baseline",
        "backgroundColor": undefined,
        "borderColor": "#195CEF",
        "borderRadius": 30,
        "borderWidth": 1,
        "height": 24,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
    testID="chip"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          [
            {
              "fontSize": 10,
              "fontWeight": "500",
              "lineHeight": 16,
              "marginLeft": 15,
              "marginRight": 15,
            },
            undefined,
            {
              "color": "#195CEF",
            },
          ]
        }
        testID="chipLabel"
      >
        Test
      </Text>
    </View>
  </View>
</View>
`;
